pkg/coredns/coredns.go

```diff
@@ -100,3 +103,37 @@ func GetUserID() int64 {
 
 	return int64(uid)
 }
+
+func DeleteCoreDNSComponents(ctx context.Context, client *kubernetes.Clientset, namespace string) error {
+	labelSelector := labels.SelectorFromSet(labels.Set{"k8s-app": "vcluster-kube-dns"}).String()
+	deleteFuncs := []func() error{
+		func() error {
+			return client.AppsV1().Deployments(namespace).DeleteCollection(ctx, metav1.DeleteOptions{}, metav1.ListOptions{LabelSelector: labelSelector})
+		},
+		func() error {
+			return client.CoreV1().Pods(namespace).DeleteCollection(ctx, metav1.DeleteOptions{}, metav1.ListOptions{LabelSelector: labelSelector})
+		},
+		func() error {
+			services, err := client.CoreV1().Services(namespace).List(ctx, metav1.ListOptions{LabelSelector: labelSelector})
+			if err != nil {
+				return err
+			}
+			if len(services.Items) != 0 {
+				for _, svc := range services.Items {
+					err = client.CoreV1().Services(namespace).Delete(ctx, svc.Name, metav1.DeleteOptions{})
+					if err != nil {
+						return err
+					}
+				}
+			}
+			return nil
+		},
+	}
+
```

Comment:
I'd rather do the best-effort cleanup by always running all delete commands and returning an error at the end:
```go
var errs []error
for _, deleteFn := range deleteFuncs {
		if err := deleteFn(); err != nil {
			errs = append(errs, err)
		}
	}
return errors.Join(errs...)
```